larq_zoo/densenet.py

```diff
@@ -1,425 +1,459 @@
-        kernel_constraint=args.constraint,
-        padding="same",
-        use_bias=False,
-    )(y)
-    return keras.layers.concatenate([x, y])
-
-
-@registry.register_model
-def binary_densenet(
-    args, input_shape, num_classes, input_tensor=None, include_top=True
-):
-    input = utils.get_input_layer(input_shape, input_tensor)
-
-    if input_shape[0] and input_shape[0] < 50:
-        x = keras.layers.Conv2D(
-            args.initial_filters,
+    input_quantizer = Field(lambda: lq.quantizers.SteSign(clip_value=1.3))
+    kernel_quantizer = Field(lambda: lq.quantizers.SteSign(clip_value=1.3))
+    kernel_constraint = Field(lambda: lq.constraints.WeightClip(clip_value=1.3))
+
+    initial_filters: int = Field(64)
+    growth_rate: int = Field(64)
+
+    # These are not `Fields`, as they should not be configurable, but set in the
+    # various concrete subclasses.
+    name: str
+    reduction: Sequence[float]
+    dilation_rate: Sequence[int]
+    layers: Sequence[float]
+    imagenet_weights_path: str
+    imagenet_no_top_weights_path: str
+
+    def densely_connected_block(self, x: tf.Tensor, dilation_rate: int = 1):
+        y = tf.keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+        y = lq.layers.QuantConv2D(
+            filters=self.growth_rate,
             kernel_size=3,
+            dilation_rate=dilation_rate,
+            input_quantizer=self.input_quantizer,
+            kernel_quantizer=self.kernel_quantizer,
+            kernel_initializer="glorot_normal",
+            kernel_constraint=self.kernel_constraint,
             padding="same",
-            kernel_initializer="he_normal",
-            use_bias=False,
-        )(input)
-    else:
-        x = keras.layers.Conv2D(
-            args.initial_filters,
-            kernel_size=7,
-            strides=2,
-            padding="same",
-            kernel_initializer="he_normal",
             use_bias=False,
-        )(input)
-
-        x = keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
-        x = keras.layers.Activation("relu")(x)
-        x = keras.layers.MaxPool2D(3, strides=2, padding="same")(x)
-
-    for block, layers_per_block in enumerate(args.layers):
-        for _ in range(layers_per_block):
-            x = densely_connected_block(x, args, args.dilation_rate[block])
-
-        if block < len(args.layers) - 1:
-            x = keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
-            if args.dilation_rate[block + 1] == 1:
-                x = keras.layers.MaxPooling2D(2, strides=2)(x)
-            x = keras.layers.Activation("relu")(x)
-            x = keras.layers.Conv2D(
-                round(x.shape.as_list()[-1] // args.reduction[block] / 32) * 32,
-                kernel_size=1,
+        )(y)
+        return tf.keras.layers.concatenate([x, y])
+
+    def build(self) -> BinaryDenseNet:
+        image_input = self.image_input
+
+        if image_input.shape[1] and image_input.shape[2] < 50:
```

Comment:
See comment in Resnet

Reply:
Resolved by db4b68c.

Reply:
Same answer as above (although I fixed a typo on this line)